qiskit/circuit/qpy_serialization.py

```diff
@@ -100,6 +100,51 @@
 by ``num_circuits`` in the file header). There is no padding between the
 circuits in the data.
 
+.. _version_4:
+
+Version 4
+=========
+
+Version 4 is identical to :ref:`version 3` except that it adds 2 new type strings
+to the INSTRUCTION_PARAM struct, ``z`` to represent ``None`` (which is encoded as
+no data), ``q`` to represent a :class:`.QuantumCircuit` (which is encoded as
+a QPY circuit), ``r`` to represent a ``range`` of integers (which is encoded as
+a :ref:`range_pack`), and ``t`` to represent a ``tuple`` of integers (which is
+encoded as npy file containing a 1-d numpy array of integers). Additionally, version
```

Comment:
In the future for control flow, we're going to need to be able to store classical expressions in the tuples as well.  Is it worth biting the bullet and making `t` a recursive type from the get-go: perhaps it's followed by a `uint32_t` of the number of elements (`uint8_t` if we want to be stingy, but there's probably no reason to bother), then that many `INSTRUCTION_PARAMS` form the tuple?

It's totally valid to defer this to later, though.

Reply:
I can do it now, I was just being lazy because I saw it was explicitly ints in the for loop params list. But it's better to try and make this future proof as it's not really any more disruptive to use a new representation or npy here.

Reply:
Actually thinking about this a bit more I'm a bit conflicted, part of me want to do this after #7582 which cleans up the parameters definitions and would make implementing this easier. But, that isn't appropriate for backporting and can't be used here. That being said if I did wait then that would mean we'd have to support `"t"` as a type string for an int array encoded as a 1-d array on V4 forever. I think I'll still eat the ugliness of implementing this now to avoid having to diverge the definitions of types long term

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/7584/commits/07da038de47de6e6c9b6a8ef36a487058b9999f8